embassy/src/util/signal.rs

```diff
@@ -67,3 +72,70 @@ impl<T: Send> Signal<T> {
         cortex_m::interrupt::free(|_| matches!(unsafe { &*self.state.get() }, State::Signaled(_)))
     }
 }
+
+struct NrWrap(u8);
+unsafe impl cortex_m::interrupt::Nr for NrWrap {
+    fn nr(&self) -> u8 {
+        self.0
+    }
+}
+
+pub struct InterruptFuture<'a, I: OwnedInterrupt> {
+    interrupt: &'a mut I,
+}
+
+impl<'a, I: OwnedInterrupt> Drop for InterruptFuture<'a, I> {
+    fn drop(&mut self) {
+        cortex_m::interrupt::free(|_| {
+            self.interrupt.remove_handler();
+            self.interrupt.disable();
+        });
+    }
+}
+
+impl<'a, I: OwnedInterrupt> InterruptFuture<'a, I> {
+    pub fn new(interrupt: &'a mut I) -> Self {
+        cortex_m::interrupt::free(|_| {
+            interrupt.set_handler(Self::interrupt_handler, ptr::null_mut());
+            interrupt.unpend();
+            interrupt.enable();
+        });
+
+        Self {
+            interrupt: interrupt,
+        }
+    }
+
+    unsafe fn interrupt_handler(ctx: *mut ()) {
+        let irq = match SCB::vect_active() {
+            scb::VectActive::Interrupt { irqn } => irqn,
+            _ => unreachable!(),
+        };
+
+        if ctx as *const _ != ptr::null() {
+            executor::raw::wake_task(ptr::NonNull::new_unchecked(ctx));
+        }
+
+        NVIC::mask(NrWrap(irq));
+    }
+}
+
+impl<'a, I: OwnedInterrupt> Future for InterruptFuture<'a, I> {
+    type Output = ();
+
+    fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        cortex_m::interrupt::free(|_| unsafe {
```

Comment:
This critical section is not needed, there's nothing that can race.

Reply:
If the interrupt fires after `is_enabled` but before `set_handler`, then Poll will return `Poll::Pending`, but the interrupt will not wake the task (because ctx will be null for that brief moment). Therefore, the critical section is required.

Reply:
Very good point. Disabling this interrupt should be enough then, instead of a full critical section that disables all interrupts.

Reply:
I'm sorry, do you have an example of a partial critical section? I haven't seen that before.

Reply:
Nope, we can't disable it because we're using the `is_enabled` as the "ready flag"...

I think calling `set_handler` *before* `is_enabled` is a good fix. If the interrupt happens between them, it'll do one "useless wake", but it will cause the future to return Ready anyway.

Reply:
Oh, I get it.

Reply:
a "partial" critical section is just disabling the interrupt, doing stuff, enabling it again :)